### PR TITLE
feat: sonicmoe nvfp4 lora merge aware

### LINKED ISSUE #3803: [ Discussion ] NVFP4 LoRA Merge
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

This is a discussion for merging LoRA to a base NVFP4 checkpoint. This is open to consider how to approach this, get feedback, and for tracking.

I think current main also discards adapter experts due to the fused nature, so see the linked PR below which solves this for both scattermoe and sonicmoe nvfp4 lora.

### ✔️ Solution

There were several ideas during discussion:
- Dequant to bf16 -> Merge and leave as bf16 checkpoint. Users can choose how they want to dequant
- Dequant to bf16 -> Merge -> Requant via model opt back to nvfp4. This is lossy. Results from #3780 show that the adapter features gets lost on requant due to 4bit.
- Dequant to bf16 -> Merge -> LLM compressor (this doesn't end up with the same nvfp4 checkpoint, but a diff 4bit compressed-tensors format)
- Do not merge and use a custom axolotl VLLM fork that supports loading adapter. Heavy maintenance. Require to add nvfp4 lora support.
- ??

One important point is that `axolotl inference ... --chat ` should work for loading adapter on the nvfp4 on the linked PR.

### ❓ Alternatives

_No response_

### 📝 Additional Context

Found another repo doing NVFP4 LoRA for the DGX Spark however, we checked their merge script and found it's not as extensive as ours and has hidden errors. Their validation does not uncover this https://github.com/waybarrios/dgx-spark-finetune-llm

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

### PR DESCRIPTION
**Pull request by NanoCode012**
# Description

Previously, the LoRA would be trained in BF16, and when merged to NVFP4 base in #3780 , it would lose its accuracy.

This PR introduces a different path, fixes #3803 , we train the LoRA fake-quantized, so that when merged, it's the same weights as in during training.

How to use: see NVFP4_LORA.qmd attached / updated example ymls. `nvfp4_merge_aware: true`

Drawback: An adapter trained this way can NOT be used loaded ontop in realtime for inference. If you want to use adapter in runtime, please use the prior training method, do not enable merge-aware.

```
┌─────────────────────────────────────────┬─────────────────────┐
  │                 Forward                 │         NLL         │
  ├─────────────────────────────────────────┼─────────────────────┤
  │ Base model                              │ 0.7164              │
  ├─────────────────────────────────────────┼─────────────────────┤
  │ Snapped Q(W_eff) = what training        │ 0.5588              │
  │ optimized = merged                      │                     │
  ├─────────────────────────────────────────┼─────────────────────┤
  │ Raw unsnapped W_eff = unmerged adapter  │ 0.7354 — worse than │
  │ at inference                            │  base               │
  └─────────────────────────────────────────┴─────────────────────┘
```

Also:
- update transformer to 5.14.1 , kernels to 0.15.x
- added support for qwen, deepseek, and other arch supported by scattermoe
- updated kernel upstream to follow latest kernel builder format

## How has this been tested?

Validated on 1xB200
- inference check
- forward pass nll check (for merge)
- cpu tests
- weights format check

## AI Usage Disclaimer

Claude

## Summary by CodeRabbit

* **New Features**
  * Added merge-aware NVFP4 MoE LoRA for SonicMoE, including bitwise-consistent baking of trained adapters into standalone NVFP4 checkpoints.
  * Added configurable merge-aware activation timing and automated adapter metadata stamping during training/saving.
  * Enhanced `merge-lora` with `--dequant` and `--override-quantizer` support, plus NVFP4 merge-aware quantization with Triton acceleration where available.
* **Documentation**
  * Added a dedicated NVFP4 MoE LoRA guide, updated navigation, support/optimization sections, and example workflows/commands.
* **Tests**
  * Added extensive merge-aware validation and correctness tests (CPU and CUDA where applicable).

### CODE REVIEW

**Comment by NanoCode012** [comment]
This test passed on the validation B200 but I suppose tooling (torch + hardware etc) causes the bits to be slightly different. Will investigate
